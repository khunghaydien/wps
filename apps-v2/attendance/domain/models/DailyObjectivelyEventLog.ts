import {
  EVENT_TYPE,
  ObjectivelyEventLogRecord as $ObjectivelyEventLogRecord,
} from './ObjectivelyEventLogRecord';
import { ObjectivelyEventLogSetting } from './ObjectivelyEventLogSetting';

export const ATTENTION_TYPE = {
  NONE: 'none',
  WARNING: 'warning',
  ERROR: 'error',
} as const;

export type AttentionType = Value<typeof ATTENTION_TYPE>;

export type ObjectivelyEventLogRecord = $ObjectivelyEventLogRecord & {
  eventLogUpdatedBy: string;
  deviatedTime: number | null;
};

export type { ObjectivelyEventLogSetting };

export type Log = {
  setting: ObjectivelyEventLogSetting;
  entering: ObjectivelyEventLogRecord & {
    eventType: typeof EVENT_TYPE.ENTERING;
  };
  leaving: ObjectivelyEventLogRecord & {
    eventType: typeof EVENT_TYPE.LEAVING;
  };
  allowingDeviationTime: number | null;
  requireDeviationReason: boolean;
};

export type DeviationReason = {
  label: string;
  value: string;
  text: string;
};

export type DailyObjectivelyEventLog = {
  id: string;
  recordId: string;
  recordDate: string;
  inpStartTime: number | null;
  inpEndTime: number | null;
  deviationReasonExtendedItemId: string;
  deviatedEnteringTimeReason: DeviationReason;
  deviatedLeavingTimeReason: DeviationReason;
  logs: [Log | null, Log | null, Log | null];
};

export type IDailyObjectivelyEventLogRepository = {
  search: (parameters: {
    employeeId?: string;
    startDate: string;
    endDate: string;
  }) => Promise<DailyObjectivelyEventLog[]>;
  saveDeviationReason: (parameters: {
    id: string;
    deviatedEnteringTimeReason: DeviationReason;
    deviatedLeavingTimeReason: DeviationReason;
    deviationReasonExtendedItemId: string;
  }) => Promise<void>;
  setToBeApplied: (parameters: {
    id: DailyObjectivelyEventLog['id'];
    records: {
      enteringId: string;
      leavingId: string;
    }[];
  }) => Promise<void>;
};

/**
 * 乖離があるかの判定をする
 *
 * 乖離があれば `true` 、なければ `false` を返す。
 * [判定]
 * - 乖離判定なし : false
 * - 連携データなし:
 *     - 入力あり : true
 *     - 入力なし : false
 * - 連携データあり
 *     - 入力なし : true
 *     - 入力あり :
 *         - 許容時間超過 : true
 *         - 許容時間以内 : false
 */
export const isOverAllowingDeviationTime = (
  inputTime: number | null,
  record: ObjectivelyEventLogRecord | null,
  allowingDeviationTime: number | null
): boolean => {
  // 乖離判定なし
  if (allowingDeviationTime === null) {
    return false;
  }
  const { time } = record;
  // 連携データなし
  if (time === null) {
    return inputTime !== null;
  }
  // 連携データあり 入力なし
  if (inputTime === null) {
    return true;
  }
  // 連携データあり 入力あり
  return allowingDeviationTime < Math.abs(inputTime - time);
};

export const getAttentionType = ({
  inputTime,
  record,
  allowingDeviationTime,
  requireDeviationReason,
  reason,
}: {
  inputTime: number | null;
  record: ObjectivelyEventLogRecord | null;
  allowingDeviationTime: number | null;
  requireDeviationReason: boolean;
  reason?: string | null;
}): AttentionType => {
  const over = isOverAllowingDeviationTime(
    inputTime,
    record,
    allowingDeviationTime
  );
  if (over) {
    // 乖離時間がある場合
    if (reason) {
      // 理由がある場合
      return ATTENTION_TYPE.NONE;
    } else {
      if (requireDeviationReason) {
        // 理由が必須場合
        return ATTENTION_TYPE.ERROR;
      } else {
        // 理由が必須ではない場合
        return ATTENTION_TYPE.WARNING;
      }
    }
  } else {
    // 乖離時間がない場合
    return ATTENTION_TYPE.NONE;
  }
};

export const getAttentionTypeAtDaily = ({
  startTime,
  endTime,
  dailyObjectivelyEventLog,
}: {
  startTime: number | null;
  endTime: number | null;
  dailyObjectivelyEventLog: DailyObjectivelyEventLog;
}): AttentionType => {
  if (!dailyObjectivelyEventLog) {
    return ATTENTION_TYPE.NONE;
  }
  let type: AttentionType = ATTENTION_TYPE.NONE;
  for (const log of dailyObjectivelyEventLog.logs) {
    if (log === null) {
      continue;
    }

    const entering = getAttentionType({
      inputTime: startTime,
      record: log.entering,
      allowingDeviationTime: log.allowingDeviationTime,
      requireDeviationReason: log.requireDeviationReason,
      reason: !dailyObjectivelyEventLog.deviatedEnteringTimeReason?.text
        ? dailyObjectivelyEventLog.deviatedEnteringTimeReason?.value
        : dailyObjectivelyEventLog.deviatedEnteringTimeReason?.text,
    });
    const leaving = getAttentionType({
      inputTime: endTime,
      record: log.leaving,
      allowingDeviationTime: log.allowingDeviationTime,
      requireDeviationReason: log.requireDeviationReason,
      reason: !dailyObjectivelyEventLog.deviatedLeavingTimeReason?.text
        ? dailyObjectivelyEventLog.deviatedLeavingTimeReason?.value
        : dailyObjectivelyEventLog.deviatedLeavingTimeReason?.text,
    });

    if (entering === ATTENTION_TYPE.ERROR || leaving === ATTENTION_TYPE.ERROR) {
      type = ATTENTION_TYPE.ERROR;
      break;
    }

    if (
      entering === ATTENTION_TYPE.WARNING ||
      leaving === ATTENTION_TYPE.WARNING
    ) {
      type = ATTENTION_TYPE.WARNING;
    }
  }

  return type;
};
