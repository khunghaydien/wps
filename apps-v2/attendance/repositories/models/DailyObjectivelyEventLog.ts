import { DailyObjectivelyEventLog } from '@attendance/domain/models/DailyObjectivelyEventLog';
import { EVENT_TYPE } from '@attendance/domain/models/ObjectivelyEventLogRecord';

export type ObjectivelyEventLogSetting = {
  id: string;
  code: string;
  name: string;
};

export type AttRecord = {
  id: string;
  recordDate: string;
  inpStartTime: number | null;
  inpEndTime: number | null;
  attSummary: {
    id: string;
    workingType: {
      objectivelyEventLogSetting1: ObjectivelyEventLogSetting | null;
      allowingDeviationTime1: number | null;
      requireDeviationReason1: boolean;
      objectivelyEventLogSetting2: ObjectivelyEventLogSetting | null;
      allowingDeviationTime2: number | null;
      requireDeviationReason2: boolean;
      objectivelyEventLogSetting3: ObjectivelyEventLogSetting | null;
      allowingDeviationTime3: number | null;
      requireDeviationReason3: boolean;
    };
  };
};

export type DailyObjectivelyEventLogRecord = {
  id: string;
  deviationReasonExtendedItemId: string;
  deviatedEnteringTimeReasonSelectedValue: string;
  deviatedEnteringTimeReasonSelectedLabel: string;
  deviatedLeavingTimeReasonSelectedValue: string;
  deviatedLeavingTimeReasonSelectedLabel: string;
  deviatedEnteringTimeReason: string;
  deviatedLeavingTimeReason: string;
  enteringEventLogId1: string;
  leavingEventLogId1: string;
  enteringTime1: number;
  leavingTime1: number;
  enteringEventLogUpdated1By: string;
  leavingEventLogUpdated1By: string;
  deviatedEnteringTime1: number;
  deviatedLeavingTime1: number;
  enteringEventLogId2: string;
  leavingEventLogId2: string;
  enteringTime2: number;
  leavingTime2: number;
  enteringEventLogUpdated2By: string;
  leavingEventLogUpdated2By: string;
  deviatedEnteringTime2: number;
  deviatedLeavingTime2: number;
  enteringEventLogId3: string;
  leavingEventLogId3: string;
  enteringTime3: number;
  leavingTime3: number;
  enteringEventLogUpdated3By: string;
  leavingEventLogUpdated3By: string;
  deviatedEnteringTime3: number;
  deviatedLeavingTime3: number;
  attRecord: AttRecord;
};

const convertRecord = (
  record: DailyObjectivelyEventLogRecord
): DailyObjectivelyEventLog => {
  const logs: DailyObjectivelyEventLog['logs'] = [
    record.attRecord.attSummary.workingType.objectivelyEventLogSetting1
      ? {
          setting: {
            id: record.attRecord.attSummary.workingType
              .objectivelyEventLogSetting1.id,
            name: record.attRecord.attSummary.workingType
              .objectivelyEventLogSetting1.name,
            code: record.attRecord.attSummary.workingType
              .objectivelyEventLogSetting1.code,
          },
          entering: {
            id: record.enteringEventLogId1,
            eventType: EVENT_TYPE.ENTERING,
            time: record.enteringTime1,
            eventLogUpdatedBy: record.enteringEventLogUpdated1By,
            deviatedTime: record.deviatedEnteringTime1,
            linked: '',
          },
          leaving: {
            id: record.leavingEventLogId1,
            eventType: EVENT_TYPE.LEAVING,
            time: record.leavingTime1,
            eventLogUpdatedBy: record.leavingEventLogUpdated1By,
            deviatedTime: record.deviatedLeavingTime1,
            linked: '',
          },
          requireDeviationReason:
            record.attRecord.attSummary.workingType.requireDeviationReason1,
          allowingDeviationTime:
            record.attRecord.attSummary.workingType.allowingDeviationTime1,
        }
      : null,
    record.attRecord.attSummary.workingType.objectivelyEventLogSetting2
      ? {
          setting: {
            id: record.attRecord.attSummary.workingType
              .objectivelyEventLogSetting2.id,
            name: record.attRecord.attSummary.workingType
              .objectivelyEventLogSetting2.name,
            code: record.attRecord.attSummary.workingType
              .objectivelyEventLogSetting2.code,
          },
          entering: {
            id: record.enteringEventLogId2,
            eventType: EVENT_TYPE.ENTERING,
            time: record.enteringTime2,
            eventLogUpdatedBy: record.enteringEventLogUpdated2By,
            deviatedTime: record.deviatedEnteringTime2,
            linked: '',
          },
          leaving: {
            id: record.leavingEventLogId2,
            eventType: EVENT_TYPE.LEAVING,
            time: record.leavingTime2,
            eventLogUpdatedBy: record.leavingEventLogUpdated2By,
            deviatedTime: record.deviatedLeavingTime2,
            linked: '',
          },
          requireDeviationReason:
            record.attRecord.attSummary.workingType.requireDeviationReason2,
          allowingDeviationTime:
            record.attRecord.attSummary.workingType.allowingDeviationTime2,
        }
      : null,
    record.attRecord.attSummary.workingType.objectivelyEventLogSetting3
      ? {
          setting: {
            id: record.attRecord.attSummary.workingType
              .objectivelyEventLogSetting3.id,
            name: record.attRecord.attSummary.workingType
              .objectivelyEventLogSetting3.name,
            code: record.attRecord.attSummary.workingType
              .objectivelyEventLogSetting3.code,
          },
          entering: {
            id: record.enteringEventLogId3,
            eventType: EVENT_TYPE.ENTERING,
            time: record.enteringTime3,
            eventLogUpdatedBy: record.enteringEventLogUpdated3By,
            deviatedTime: record.deviatedEnteringTime3,
            linked: '',
          },
          leaving: {
            id: record.leavingEventLogId3,
            eventType: EVENT_TYPE.LEAVING,
            time: record.leavingTime3,
            eventLogUpdatedBy: record.leavingEventLogUpdated3By,
            deviatedTime: record.deviatedLeavingTime3,
            linked: '',
          },
          requireDeviationReason:
            record.attRecord.attSummary.workingType.requireDeviationReason3,
          allowingDeviationTime:
            record.attRecord.attSummary.workingType.allowingDeviationTime3,
        }
      : null,
  ];

  return {
    id: record.id,
    recordId: record.attRecord.id,
    recordDate: record.attRecord.recordDate,
    inpStartTime: record.attRecord.inpStartTime,
    inpEndTime: record.attRecord.inpEndTime,
    deviationReasonExtendedItemId: record.deviationReasonExtendedItemId,
    deviatedEnteringTimeReason: {
      label: record.deviatedEnteringTimeReasonSelectedLabel,
      value: record.deviatedEnteringTimeReasonSelectedValue,
      text: record.deviatedEnteringTimeReason,
    },
    deviatedLeavingTimeReason: {
      label: record.deviatedLeavingTimeReasonSelectedLabel,
      value: record.deviatedLeavingTimeReasonSelectedValue,
      text: record.deviatedLeavingTimeReason,
    },
    logs,
  };
};

export const convert = (
  dailyRecordList: DailyObjectivelyEventLogRecord[]
): DailyObjectivelyEventLog[] =>
  dailyRecordList ? dailyRecordList.map(convertRecord) : null;
