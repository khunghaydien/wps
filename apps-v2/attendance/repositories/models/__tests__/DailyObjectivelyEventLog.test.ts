import cloneDeep from 'lodash/cloneDeep';

import {
  convert,
  DailyObjectivelyEventLogRecord,
} from '../DailyObjectivelyEventLog';
import { defaultValue } from './mocks/DailyObjectivelyEventLog.mock';

describe('convert', () => {
  it('should do', async () => {
    const $defaultValue: DailyObjectivelyEventLogRecord[] =
      cloneDeep(defaultValue);
    // Overwrite boolean values because it can't be checked that converting is correct.
    $defaultValue[0].attRecord.attSummary.workingType.requireDeviationReason1 =
      'requireDeviationReason1' as unknown as boolean;
    $defaultValue[0].attRecord.attSummary.workingType.requireDeviationReason2 =
      'requireDeviationReason2' as unknown as boolean;
    $defaultValue[0].attRecord.attSummary.workingType.requireDeviationReason3 =
      'requireDeviationReason3' as unknown as boolean;
    expect(convert($defaultValue)).toMatchSnapshot();
  });
});
