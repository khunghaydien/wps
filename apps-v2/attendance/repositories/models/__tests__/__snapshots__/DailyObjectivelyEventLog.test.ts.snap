// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convert should do 1`] = `
Array [
  Object {
    "deviatedEnteringTimeReason": Object {
      "label": "deviatedEnteringTimeReasonSelectedLabel",
      "text": "Deviated Entering Time Reason",
      "value": "deviatedEnteringTimeReasonSelectedValue",
    },
    "deviatedLeavingTimeReason": Object {
      "label": "deviatedLeavingTimeReasonSelectedLabel",
      "text": "Deviated Leaving Time Reason",
      "value": "deviatedLeavingTimeReasonSelectedValue",
    },
    "deviationReasonExtendedItemId": "deviationReasonExtendedItemId",
    "id": "dailyObjectivelyEventLogId",
    "inpEndTime": 1075,
    "inpStartTime": 415,
    "logs": Array [
      Object {
        "allowingDeviationTime": null,
        "entering": Object {
          "deviatedTime": 5,
          "eventLogUpdatedBy": "Entering event log updated1 by",
          "eventType": "Entering",
          "id": "enteringEventLogId1",
          "linked": "",
          "time": 480,
        },
        "leaving": Object {
          "deviatedTime": 6,
          "eventLogUpdatedBy": "Leaving event log updated1 by",
          "eventType": "Leaving",
          "id": "leavingEventLogId1",
          "linked": "",
          "time": 1020,
        },
        "requireDeviationReason": "requireDeviationReason1",
        "setting": Object {
          "code": "CODE_NAME_1",
          "id": "setting1",
          "name": "name1",
        },
      },
      Object {
        "allowingDeviationTime": 5,
        "entering": Object {
          "deviatedTime": 10,
          "eventLogUpdatedBy": "Entering event log updated2 by",
          "eventType": "Entering",
          "id": "enteringEventLogId2",
          "linked": "",
          "time": 470,
        },
        "leaving": Object {
          "deviatedTime": 9,
          "eventLogUpdatedBy": "Leaving event log updated2 by",
          "eventType": "Leaving",
          "id": "leavingEventLogId2",
          "linked": "",
          "time": 1011,
        },
        "requireDeviationReason": "requireDeviationReason2",
        "setting": Object {
          "code": "CODE_NAME_2",
          "id": "setting2",
          "name": "name2",
        },
      },
      Object {
        "allowingDeviationTime": 60,
        "entering": Object {
          "deviatedTime": 1,
          "eventLogUpdatedBy": "Entering event log updated3 by",
          "eventType": "Entering",
          "id": "enteringEventLogId3",
          "linked": "",
          "time": 481,
        },
        "leaving": Object {
          "deviatedTime": 3,
          "eventLogUpdatedBy": "Leaving event log updated3 by",
          "eventType": "Leaving",
          "id": "leavingEventLogId3",
          "linked": "",
          "time": 1023,
        },
        "requireDeviationReason": "requireDeviationReason3",
        "setting": Object {
          "code": "CODE_NAME_3",
          "id": "setting3",
          "name": "name3",
        },
      },
    ],
    "recordDate": "2020-01-01",
    "recordId": "recordId",
  },
]
`;
